api/api_full.go

```diff
@@ -936,6 +956,39 @@ type FullNode interface {
 	// F3IsRunning returns true if the F3 instance is running, false if it's not running but
 	// it's enabled, and an error when disabled entirely.
 	F3IsRunning(ctx context.Context) (bool, error) //perm:read
+	// F3GetProgress returns the progress of the current F3 instance in terms of instance ID, round and phase.
+	F3GetProgress(ctx context.Context) (F3Progress, error) //perm:read
+}
+
+// F3ParticipationToken represents a binary token that authorizes a miner to
+// participate in the F3 consensus.
+type F3ParticipationToken []byte
+
+// F3ParticipationLease defines the lease granted to a storage provider for
+// participating in F3 consensus, detailing the session identifier, issuer,
+// subject, and the expiration instance.
+type F3ParticipationLease struct {
+	// Session uniquely identifies the session in which this lease is issued. This
+	// information can then be used to infer if the issuer of the token has restarted
+	// since the lease was issued.
+	Session uuid.UUID
+	// Issuer is the identity of the node that issued the lease.
+	Issuer peer.ID
+	// Subject is the actor ID of the lease holder.
+	Subject uint64
+	// ExpireAfterInstance is the GPBFT instance ID, beyond which this lease expires.
+	// See F3Progress.
+	ExpireAfterInstance uint64
```

Comment:
I _think_ we need a "not before" as well, right? I.e., we need to grant a lease between instances A and B.

Reply:
I had that initially but removed it for simplicity. Without it I think the design would work correctly and miners renew within the last 2 instances.

Without a strict boundary i think the token design as is would still work correctly. Because, each node will only give one lease out per miner, and would only renew if the old one is expired or was issued by itself.

With it the API is more defensive against miners that call to get a token too often, but the entire implementation here is based on the assumption that miner and node are trusted.

Right?

Reply:
My concern is that "expired" is relative. E.g.:

1. Miner sees that the latest instance is X according to lotus A.
2. Miner gives a lease to lotus B, but lotus B thinks that the latest instance is X-1.

Reply:
Got it thanks

I think this won't cause problems because the miners renew lease within the last 2 instances remaining in the lease. But that's not enforced by the Lotus node API. 

I'll enforce it 👍

Reply:
Oh, I see. IMO, this API is backwards: the _Miner_ should be the one in control, not the node. 

That is:

1. Miner calls "pre-participate" and gets a peer id, session id (unnecessary?), and the node's current instance.
2. Miner calls "participate" with a not-before instance, not-after instance, and a peer ID.

Importantly, the Miner decides the instance range to lease out, not the node.

Reply:
Hm. Well, as long as the "not before" is encoded in the ticket this should be fine. But I still think the miner should be able to specify the "not before" time when getting the ticket, at least.

Reply:
> the Miner should be the one in control, not the node.

Why?  I mean that's valid too but why is that better than node deciding the lease range?

Reply:
But in that case, you're right. Given that this is the lease as specified by the _node_, there's no reason to have a "not before".

Reply:
> miner should be able to specify the "not before"

Thinking about it more, I agree we should add this. Because, the F3 progress across nodes may not be exactly the same. I think i also need to encode currentInstance in the ticket response in couple of places.


While we are on the subject, do we care about the session at all? Right now it is a noop apart from the informative log line to indicate that the node must have restarted.